### DISCUSSION TITLE
Group updates to have consistent versions

### DISCUSSION THREAD

**Discussion by colin-janke** [renovatebot/renovate#15923] [Suggest an Idea]
Something of a speculative idea:

We have a number of dependencies which share a common root dependency. Eg:

- My project depends on projects A B and C
- Project A also depends on project C
- Project B also depends on project C

We use a combination of package rules (groupings), the enforcer plugin and mvn dependency exclusions to try and control what version of different projects we use and to try and keep things consistent.  However, this is getting more complex, can give surprising errors when the exclusions cause us to use incompatible code and can make the grouped MR take a long time to get merged as we get renovate-generated patch releases of the low-level library while trying to assemble things.

For example, this can result in a case where the latest version of C is 1.2 but the latest versions of A and B depend on C v1.1. We then get a renovate MR which tries to update to the latest versions of A, B and C, which then fails the enforcer plugin checks.

What I think would be nice would be some extension of the grouping rules which would allow renovate to generate compatible sets of versions. So if the latest version of C is 1.2 but the latest release of A and B still depend on C v 1.1, we would get a renovate MR bringing us up to the latest versions of A and B with C at v1.1. This could then automerge and would keep us generally closer to latest on all of these dependencies as we wouldn't have so many failed MRs.

What I think would be additionally useful would be some kind of indication of what is being blocked.  Maybe in the form of an additional MR with the latest version of C so we can see that we need to get a version of A and B to depend on it.

**Comment by rarkins** [comment]
One challenge is that if you intentionally filter/withhold updates because some package doesn't support it yet, you could get stuck forever *and* be unaware of it. i.e. you may not even know that 1.2 exists and that you're missing it, if Renovate only proposes 1.1. In fact it could be up to 1.8 and you're a year behind and you still don't notice, because you don't get any PR.

Due to this today we err on the side of visibility (preferring to raise a PR and you can see there's incompatible versions) than withholding, but ideally we'd find a way to achieve both.

**Reply by colin-janke** [reply]
I agree.  that was exactly the thought behind my last paragraph. In my example, we currently get:

* A v latest B v latest C v latest -> fails enforcer

This leads to us sitting on an older version of all three libraries for a time until we manually update or the versions finally align.

I would suggest (but am happy to hear other ideas)

* A v latest B v latest C v1.1 -> passes and automerges
* A v latest B v latest C v latest -> C v latest (once the first has merged in, the A and B changes will not be required) -> fails enforcer

By creating _both_ MRs, we get visibility of the later version of C while still auto-merging A, B and a less out-of-date version of C.

**Reply by rarkins** [reply]
The next challenge is "how does Renovate know to create both MRs?"

If it's done via static configuration then fine - that's normal. But then you may need to wait for the option `separateMultipleMinor` and then set it for every package which *might* need it.

**Reply by colin-janke** [reply]
I don't know anything about how renovate works under the hood, so I am not expecting this to happen quickly, even if you do accept the idea and schedule it in.

I was imagining something like:

```
"groupName": "myMessyDependencies",
      "packagePatterns": ["^com\\.me\\.A", "^com\\.me\\.B", "^com\\.me\\.C"],
      "enabled": "true"
      "gatherCompatibleVersions": "true"
```

if that is an answer to your question? I was anticipating having to set a flag on the group definition.

`separateMultipleMinor` sounds from thee name like it might do close to what we want.  Do you know how it would interact with groups like this? would it just create n! MRs across each combination of versions and close the ones which become obsolete? That sounds like it would work, though it would generate a lot of MRs in our case.

**Reply by rarkins** [reply]
I think we could call this new option `versionConsistency` and have possible values `none` (existing/default), `patch`, `minor`, `major`. If it's set to a non-none value then it means Renovate should try to "lower" the update to find a shared version, and if possible it will be considered a failed internal status (like stability days). The user would need to guide Renovate as to whether each update in the PR needs the exact same version (e.g. `patch`) or if minor or major compatibility are enough.

This would *not* mean that Renovate is doing any transitive logic (e.g knowing that A depends on B and C in your example) but simply that it's trying to align the versions of A, B and C if told to.

Would this satisfy your requirements?

**Reply by colin-janke** [reply]
"This would not mean that Renovate is doing any transitive logic (e.g knowing that A depends on B and C in your example) but simply that it's trying to align the versions of A, B and C if told to." I don't understand this statement.  How would it find a compatible version without looking at the dependencies of the dependencies?

**Reply by colin-janke** [reply]
It sounds very good in principal.

**Reply by rarkins** [reply]
I thought A, B and C were all related libraries with related versioning. If that's not the case, then this won't satisfy your requirements, and at this point I don't think we have any way to, without either of these bad options:
- Reverse engineering the transitive package resolution algorithm of every manager (they all do it differently), or
- Brute forcing it by first seeing if the install works, then trying earlier versions/combinations, etc.

**Reply by colin-janke** [reply]
Ah, no, they do not have related versions.

I guess I would expect that modules with synched versions would have synched releases, which should eliminate this issue, but I guess that might not be the case for everyone.

I have another awful idea - run the enforcer plugin and try to parse the failure output. :)

Could you inspect the dependencies one step down rather than fully recursive? That would help us, but I don't know if it would be any easier.

As I say, I don't have any knowledge of the internals of renovate, so I can't suggest anything useful. If you (collectively?) can figure out a way of doing this, I think it would be great. If not, we'll have to continue as we are/see if there are any other solutions.

**Reply by rarkins** [reply]
I would *like* to solve this problem, but it seems pretty hard. And as you may guess, our internals certainly don't assist with it, because "artifact updating" (the point where we would learn if there's a problem) is essentially the last step in our logic. By that point we've already decided which packages to group together, which versions, made updates to the manifest files, and essentially decided on the release notes/PR contents.

**Reply by esafak** [reply]
I think we can overcome this problem if we leverage the resolver:

**1. Type the outcome + make it queryable.** Today every resolver failure is one opaque `artifactError`. Classify it (`resolved-clean | duplicates | resolution-conflict | env-error`) and parse the *blocker's constraint* out of output the resolver already prints (cargo's `required by B`, npm `ERESOLVE`, bundler, composer, pub…). Now "B can't accept C 2.0" is distinct from "registry was down," and the blocker set becomes the state later phases key off.

**2. Resolution-guided backtracking.** When the typed conflict says "C → 2.0 blocked because B requires ^1," you know the blocker's exact constraint, hence the compatible ceiling on C. So:

1. Compute the latest C that satisfies B every run, and surface blocks in the dashboard.
2. Re-run the resolver against that target.
3. If it still conflicts, repeat with the next-lower blocker — a bounded loop until resolution succeeds or you hit the current version.
4. Open a PR if the resolved set is better than current one.

I would have commented on https://github.com/renovatebot/renovate/discussions/17815 but it is locked.
